File: app/src/main/java/com/example/kursach/data/repository/AuthRepositoryImpl.kt
package com.example.kursach.data.repository

import com.example.kursach.core.di.IoDispatcher
import com.example.kursach.core.storage.SessionStorage
import com.example.kursach.data.remote.KursachApi
import com.example.kursach.data.remote.dto.LoginRequestDto
import com.example.kursach.data.remote.dto.RegisterRequestDto
import com.example.kursach.data.remote.mappers.toDomain
import com.example.kursach.domain.model.UserProfile
import com.example.kursach.domain.repository.AuthRepository
import com.example.kursach.domain.usecase.RegistrationData
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext
import retrofit2.HttpException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepositoryImpl @Inject constructor(
    private val api: KursachApi,
    private val sessionStorage: SessionStorage,
    @IoDispatcher private val ioDispatcher: CoroutineDispatcher,
) : AuthRepository {

    override suspend fun login(email: String, password: String): Result<UserProfile> =
        withContext(ioDispatcher) {
            runCatching {
                val tokenResponse = api.login(LoginRequestDto(email = email, password = password))
                sessionStorage.saveToken(tokenResponse.accessToken)
                api.fetchCurrentUser().toDomain()
            }.onFailure {
                sessionStorage.clear()
            }
        }

    override suspend fun register(data: RegistrationData): Result<UserProfile> =
        withContext(ioDispatcher) {
            runCatching {
                api.register(
                    RegisterRequestDto(
                        email = data.email,
                        password = data.password,
                        firstName = data.firstName,
                        lastName = data.lastName,
                        birthDate = data.birthDate,
                        region = data.region,
                        city = data.city,
                    ),
                )
                val tokenResponse = api.login(LoginRequestDto(email = data.email, password = data.password))
                sessionStorage.saveToken(tokenResponse.accessToken)
                api.fetchCurrentUser().toDomain()
            }.onFailure {
                sessionStorage.clear()
            }
        }

    override suspend fun fetchProfile(forceRefresh: Boolean): Result<UserProfile> =
        withContext(ioDispatcher) {
            val token = sessionStorage.authToken.value
            if (token.isNullOrBlank()) {
                return@withContext Result.failure(IllegalStateException(ERROR_NO_SESSION))
            }
            runCatching {
                api.fetchCurrentUser().toDomain()
            }.onFailure { error ->
                if (error is HttpException && error.code() == 401) {
                    sessionStorage.clear()
                }
            }
        }

    override suspend fun logout(): Result<Unit> =
        withContext(ioDispatcher) {
            runCatching {
                api.logout()
                sessionStorage.clear()
            }.onFailure {
                sessionStorage.clear()
            }
        }
}

private const val ERROR_NO_SESSION = "\u0421\u0435\u0441\u0441\u0438\u044f \u043d\u0435 \u043d\u0430\u0439\u0434\u0435\u043d\u0430"

File: lib/src/data/repositories/auth_repository_impl.dart
import '../../domain/entities/user_profile.dart';
import '../../domain/repositories/auth_repository.dart';
import '../../domain/usecases/registration_data.dart';
import '../../domain/usecases/profile_update_data.dart';
import '../mappers/user_mapper.dart';
import '../services/kursach_api.dart';
import '../../core/storage/auth_token_storage.dart';

class AuthRepositoryImpl implements AuthRepository {
  AuthRepositoryImpl(this._api, this._tokenStorage);

  final KursachApi _api;
  final AuthTokenStorage _tokenStorage;

  @override
  Future<UserProfile> login({
    required String email,
    required String password,
  }) async {
    try {
      final tokenResponse = await _api.login(email: email, password: password);
      await _tokenStorage.saveToken(tokenResponse.accessToken);
      final user = await _api.fetchCurrentUser();
      return user.toDomain();
    } catch (error) {
      await _tokenStorage.clear();
      rethrow;
    }
  }

  @override
  Future<UserProfile> register(RegistrationData data) async {
    try {
      await _api.register(
        email: data.email,
        password: data.password,
        firstName: data.firstName,
        lastName: data.lastName,
        birthDate: data.birthDate,
      );
      final tokenResponse = await _api.login(
        email: data.email,
        password: data.password,
      );
      await _tokenStorage.saveToken(tokenResponse.accessToken);
      final user = await _api.fetchCurrentUser();
      return user.toDomain();
    } catch (error) {
      await _tokenStorage.clear();
      rethrow;
    }
  }

  @override
  Future<UserProfile> fetchProfile({bool forceRefresh = false}) async {
    final token = await _tokenStorage.readToken();
    if (token == null || token.isEmpty) {
      throw const SessionMissingException();
    }
    try {
      final dto = await _api.fetchCurrentUser();
      return dto.toDomain();
    } catch (error) {
      await _tokenStorage.clear();
      rethrow;
    }
  }

  @override
  Future<void> logout() async {
    try {
      await _api.logout();
    } finally {
      await _tokenStorage.clear();
    }
  }

  @override
  Future<UserProfile> updateProfile(ProfileUpdateData data) async {
    final dto = await _api.updateProfile(
      email: data.email,
      firstName: data.firstName,
      lastName: data.lastName,
      password: data.password,
    );
    return dto.toDomain();
  }
}

class SessionMissingException implements Exception {
  const SessionMissingException();

  @override
  String toString() => 'SessionMissingException: Сессия не найдена';
}
